reference-implementation/lib/readable-byte-stream.js

```diff
@@ -0,0 +1,965 @@
+      }
+
+      RespondToReadIntoRequestsFromQueue(this, reader);
+
+      return;
+    }
+
+    CallPullOrPullIntoLaterIfNeeded(this);
+  }
+}
+
+class ReadableByteStreamReader {
+  constructor(stream) {
+    if (!IsReadableByteStream(stream)) {
+      throw new TypeError('ReadableByteStreamReader can only be constructed with a ReadableByteStream instance');
+    }
+    if (IsReadableByteStreamLocked(stream)) {
+      throw new TypeError('This stream has already been locked for exclusive reading by another reader');
+    }
+
+    InitializeReadableByteStreamReader(this, stream);
+
+    this._readRequests = [];
+  }
+
+  get closed() {
+    if (!IsReadableByteStreamReader(this)) {
+      return Promise.reject(
+        new TypeError('ReadableByteStreamReader.prototype.closed can only be used on a ReadableByteStreamReader'));
+    }
+
+    return this._closedPromise;
+  }
+
+  cancel(reason) {
+    if (!IsReadableByteStreamReader(this)) {
+      return Promise.reject(
+        new TypeError('ReadableByteStreamReader.prototype.cancel can only be used on a ReadableByteStreamReader'));
+    }
+
+    if (this._state === 'closed') {
+      return Promise.resolve(undefined);
+    }
+
+    if (this._state === 'errored') {
+      return Promise.reject(this._storedError);
+    }
+
+    assert(this._ownerReadableByteStream !== undefined, 'This reader must be attached to a stream');
+
+    return CancelReadableByteStream(this._ownerReadableByteStream, reason);
+  }
+
+  read() {
+    if (!IsReadableByteStreamReader(this)) {
+      return Promise.reject(
+        new TypeError('ReadableByteStreamReader.prototype.read can only be used on a ReadableByteStreamReader'));
+    }
+
+    if (this._state === 'closed') {
+      return Promise.resolve(CreateIterResultObject(undefined, true));
+    }
+
+    if (this._state === 'errored') {
+      return Promise.reject(this._storedError);
+    }
+
+    assert(this._ownerReadableByteStream !== undefined, 'This reader must be attached to a stream');
+    assert(this._ownerReadableByteStream._state === 'readable', 'The owner stream must be in readable state');
+
+    const promise = new Promise((resolve, reject) => {
+      this._readRequests.push({resolve, reject});
+    });
+
+    PullFromReadableByteStream(this._ownerReadableByteStream);
+
+    return promise;
+  }
+
+  releaseLock() {
+    if (!IsReadableByteStreamReader(this)) {
+      throw new TypeError(
+          'ReadableByteStreamReader.prototype.releaseLock can only be used on a ReadableByteStreamReader');
+    }
+
+    if (this._ownerReadableByteStream === undefined) {
+      return;
+    }
+
+    if (this._readRequests.length > 0) {
+      throw new TypeError('Tried to release a reader lock when that reader has pending read() calls un-settled');
+    }
+
+    if (this._ownerReadableByteStream._state === 'errored') {
+      MarkReadableByteStreamReaderErrored(this, this._ownerReadableByteStream._storedError)
```

Comment:
Since we have auto release, actually this line is unreached. I just removed this and replaced the if with assertion. f3be413272042e204083abbf52aa875e9cd753bc